Sources/AblyChat/RoomLifecycleManager.swift

```diff
@@ -0,0 +1,297 @@
+                        // TODO: The spec assumes this will be populated, but working in a multi-threaded environment means it might not be (https://github.com/ably-labs/ably-chat-swift/issues/49)
+                        preconditionFailure("Contributor entered FAILED but its errorReason is not set")
+                    }
+
+                    let error = ARTErrorInfo(chatError: .detachmentFailed(feature: contributor.feature, underlyingError: contributorError))
+
+                    if firstDetachError == nil {
+                        // We’ll throw this after we’ve tried detaching all the channels
+                        firstDetachError = error
+                    }
+
+                    // This check is CHA-RL2h2
+                    if current != .failed {
+                        changeStatus(to: .failed, error: error)
+                    }
+                default:
+                    // CHA-RL2h3: Retry until detach succeeds, with a pause before each attempt
+                    while true {
+                        do {
+                            logger.log(message: "Will attempt to detach non-failed contributor \(contributor) in 1s.", level: .info)
+                            // TODO: what's the correct wait time? (https://github.com/ably/specification/pull/200#discussion_r1763799223)
+                            try await clock.sleep(timeInterval: 1)
+                            logger.log(message: "Detaching non-failed contributor \(contributor)", level: .info)
+                            try await contributor.channel.detach()
+                            break
+                        } catch {
+                            // Loop repeats
+                            logger.log(message: "Failed to detach non-failed contributor \(contributor), error \(error). Will retry.", level: .info)
+                        }
+                    }
+                }
+            }
+        }
+
+        if let firstDetachError {
+            // CHA-RL2f
+            throw firstDetachError
+        }
+
+        // CHA-RL2g
+        changeStatus(to: .detached)
+    }
+
+    /// Implementes CHA-RL3’s RELEASE operation.
+    internal func performReleaseOperation() async {
+        switch current {
+        case .released:
+            // CHA-RL3a
+            return
+        case .detached:
+            // CHA-RL3b
+            changeStatus(to: .released)
+            return
+        case .releasing, .initialized, .attached, .attaching, .detaching, .suspended, .failed:
+            break
+        }
+
+        changeStatus(to: .releasing)
+
+        // CHA-RL3d
+        for contributor in contributors {
+            detachLoop: while true {
+                let contributorState = await contributor.channel.state
+
+                // CHA-RL3e
+                guard contributorState != .failed else {
+                    logger.log(message: "Contributor \(contributor) is FAILED; skipping detach", level: .info)
+                    break
```

Comment:
should this be `break detachLoop`?

Reply:
No; this one is to break the `for contributor in contributors`; i.e. skip that contributor.

Reply:
Wait, hang on

Reply:
I think I’m talking nonsense, will take a look

Reply:
the log line above implies you want to skip detach, but the break would only exit the guard. It'd still try to detach a couple lines below i think

Reply:
`break` and `break detachLoop` are equivalent, since `break` breaks the innermost loop, which is `detachLoop` anyway. So the code works as written (I was confused, because it looked odd but the test that checks for exactly this was passing), but will get rid of `detachLoop` label

Reply:
Done

Reply:
ahh cool! that makes sense - the two different `break`'s confused me too